UWV Open Match data

galamit86 opened the pull request:
# This PR:
* Updates:
  * `register_zip_csv` so it:
    * Can be mapped on multiple urls, each containing a zip file
    * Can process an arbitrarily deep folder structure in the zip file, **without** maintaing that structure (flattens all csvs to one list).
    * Ends with files on GCS, where previously also created the BQ dataset.
    * Takes `enconding` as a parameter, and hands it over to `pyarrow.csv.ReadOptions`
  * `tasks` module, accordingly.
* Adds:
  * `run_uwv`:
    * Uploading all UWV Open Match data from from overheid.nl to GCS, and linking in BQ.
  * `ckan` module - a starting point to aggregate different functions interacting with the CKAN api.
__________
# Next steps:
* Decide (and implement) how to handle the existence of text files along side the csv files in the UWV Open Match Data:
  * **Option 1**: Create a `zip_txt`(/`zip_fwf`) flow, and add it to the `run_uwv` using the same URLs. PRO: simplest. CON: Not really DRY, as a large part will look exactly the same as `zip_csv`.
  * **Option 2**: Enhance the current `zip_csv` flow to handle different types of files differently. PRO: DRY. CON: More complicated to implement, might be too much work for a situation I am not sure that happens a lot.
  * **Option 3**: Separate the unzipping and conversion to parquet into separate flows. PRO: DRY and relatively simple to build. CON: Feels like it breaks the idea of a `flow` into too small a slice - finishing the flow with files "halfway through" on a local machine - not ideal when thinking of scaling and parallelism.
______
# Known issues:
* To separate `something_to_gcs` and `gcs_to_bq`, `run_uwv` uses a prefect's concept of [flow-of-flows](https://docs.prefect.io/core/idioms/flow-to-flow.html). This works by only scheduling the next flow after the upstream flow concluded successfully. There are 2 issues with this approach:
  * Since the dependency is about the **scheduling** of the 2nd flow, the process running the `run_flow` should be kept awake until the 1st flow finishes. Considering we envision running the `run_flows` locally and quickly, this does not fit the bill. Have not found a good solution yet.
  * If the first flow is not successful, we might still want to run the second flow - to be considered.
* The `zip_csv` flow cannot handle different datasets at once (as 1 dataset -> 1 GCS folder) - this should be done by running multiple flows with different parameters.

dkapitan commented:
Lgtm. Regarding your options, option 3 sounds best to me. Don't understand why that breaks the idea of a flow. My thought is that you would branch within a flow, one branch processing the fetched zip's to parquet, the other one with csv

galamit86 commented:
@dkapitan 
Updated the code to Implement a switch logic inside the flow to handle `csv` or `txt` files differently (`tasks.csv_to_parquet` or `tasks.fwf_to_json`). It's not ideal, but works for now, until we run into some new use cases. 

Currently, I renamed the flow from `zip_csv` to `zip_file`, and it:

1. Downloads a zip file from a url.
2. Unzips it, maintaining the folder structure.
3. Checks all files, an handles:
  3.1. `csv` files -> converts to `parquet`
  3.2. `txt` files -> assumes they are `fwf` files, and converts to `ndjson` (this is my least favourite part)
4. Uploads to a specific GCS folder, maintaining the same folder structure within it.

In `run_uwv`, that flow is launched, and after it is finished, `gcs_folder_to_bq_flow` is launched, creating the dataset and external tables. This flow was also updated (all the way down through `tasks.gcs_folder_to_bq` and `utils.create_linked_tables`), to be able to differentiate `parquet` files from `json` files, as they require a different configuration. For the `json` files, using Google's autodetection works great, and solved the type inference issues. They do not offer that for `parquet` files, however and I'm looking into that - currently it looks the easiest way is to go through `pandas` - but I'll leave that for a new issue/branch.

dkapitan reviewed (commented):
lgtm
